### PR TITLE
[12.0 ][MIG] base_location: Migration to 12.0

### PR DESCRIPTION
**Pull request by aitorbouzas**
**Refactored**

This module has now been refactored to be more consistent with what base_address_city offers to the location management.
Now, every res.city record has a relation One2many to res.city.zip (old res.better.zip). This way, every zip has a realted city too.
Zips can be searched through city code, zip or city name (same as before).
Modified tests and deleted not needed tests.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
@aitorbouzas I'm tomorrow on holidays, but on Wednesday, we can meet for seeing the problem here + the refactoring that this module needs IMO since the introduction of `base_address_city`.

File: base_location/README.rst
Comment (nikul-serpentcs) [inline base_location/README.rst]: @aitorbouzas IMO Change target `12.0`

File: base_location/__manifest__.py
Comment (nikul-serpentcs) [inline base_location/__manifest__.py]: Replace URL as per [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L10)

File: base_location/__manifest__.py
Comment (nikul-serpentcs) [inline base_location/__manifest__.py]: Follow this sequence https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L27

**Review by nikul-serpentcs — COMMENTED** [review]
Improve Code

**Comment by aitorbouzas** [comment]
@pedrobaeza Ok Pedro, didn't bear in mind the new module when migrating this one. Just write to me through email and we'll fix an hour to meet.

**Comment by aitorbouzas** [comment]
This module has now been refactored. Now it works more consistently with ```base_address_city``` (check commit info, PR's description or files changes for more info).

File: base_location/models/res_city.py
Comment (pedrobaeza) [inline base_location/models/res_city.py]: Do you think this code brings too much value? Are there official codes for the cities?

File: base_location/models/res_city_zip.py
Comment (pedrobaeza) [inline base_location/models/res_city_zip.py]: ResCityZip

File: base_location/models/res_city_zip.py
Comment (pedrobaeza) [inline base_location/models/res_city_zip.py]: This would point only to a point, not defining an area. Do you think this adds enough value? I know they were previously on the module, but I doubt this has been useful any time.

File: base_location/models/res_city_zip.py
Comment (pedrobaeza) [inline base_location/models/res_city_zip.py]: Isn't easier to just do `name.append(rec.name)` (it's already a string)?

File: base_location/models/res_city_zip.py
Comment (pedrobaeza) [inline base_location/models/res_city_zip.py]: I think it should be required

File: base_location/models/res_city_zip.py
Comment (pedrobaeza) [inline base_location/models/res_city_zip.py]: Remove this `if` being the name required

File: base_location/models/res_city_zip.py
Comment (pedrobaeza) [inline base_location/models/res_city_zip.py]: `name.append(rec.city_id.name)`

File: base_location/models/res_city_zip.py
Comment (pedrobaeza) [inline base_location/models/res_city_zip.py]: `name.append(rec.city_id.state_id.name)`

File: base_location/models/res_city_zip.py
Comment (pedrobaeza) [inline base_location/models/res_city_zip.py]: `name.append(rec.city_id.country_id.name)`

File: base_location/models/res_city_zip.py
Comment (pedrobaeza) [inline base_location/models/res_city_zip.py]: I'm thinking that maybe you can add a computed stored field depending on these values for not overloading each `name_get` call, having to access 3 tables. This will rise the entries addition (for example, through `base_location_geonames_import`), but will speed up location search, which is the most habitual task. Being called `display_name`, it will be automatically done without overwriting this method.

File: base_location/models/res_city_zip.py
Comment (pedrobaeza) [inline base_location/models/res_city_zip.py]: I think changing `_rec_name` to `display_name` will perform the search over the new computed field, handling all these cases, and making it more performant.

File: base_location/models/res_company.py
Comment (pedrobaeza) [inline base_location/models/res_company.py]: On v11, I remember this to be problematic, not filling correctly fields, but as it's only one time operation, I ignored the problem. Have you checked if this works?

File: base_location/models/res_partner.py
Comment (pedrobaeza) [inline base_location/models/res_partner.py]: Only reset values of state and country if they are filled on the city.

File: base_location/models/res_partner.py
Comment (pedrobaeza) [inline base_location/models/res_partner.py]: There are other onchange for this field on the bottom.

File: base_location/readme/CONFIGURE.rst
Comment (pedrobaeza) [inline base_location/readme/CONFIGURE.rst]: This menu no longer exists on v12, and I think that we should put Locations menu now on *Contacts > Configuration > Localization*, as we have that menu path for the rest of the localization elements (States, Countries...), and thus to add `contacts` as dependency, which is very little trade-off.

File: base_location/readme/CONTRIBUTORS.rst
Comment (pedrobaeza) [inline base_location/readme/CONTRIBUTORS.rst]: Please remove my old email

File: base_location/security/ir.model.access.csv
Comment (pedrobaeza) [inline base_location/security/ir.model.access.csv]: Not sure if we should grant read access to portal/website users as well.

File: base_location/views/res_city.xml
Comment (pedrobaeza) [inline base_location/views/res_city.xml]: Please don't replace, but hide the element, or we expose to a possible error if other module takes this element as reference.

File: base_location/views/res_city_zip.xml
Comment (pedrobaeza) [inline base_location/views/res_city_zip.xml]: Better to do this editable

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Thank you very much for the changes! Excellent work

@rvalyi you'll be glad to know that with this, we normalize city DB structure, having only one record per city.

Several things to do for making it perfect:

-  ~100% coverage. It's a pity that some cases are not covered now while they were previously. 
- Views files are named as `model_views.xml`. This is for differentiating them from for example other files that adds data for that model when looking by file name.

File: base_location/models/res_city_zip.py
Comment (pedrobaeza) [inline base_location/models/res_city_zip.py:17]: Add an SQL constraint with `UNIQUE(city_id, name)`

File: base_location/models/res_city.py
Comment (pedrobaeza) [inline base_location/models/res_city.py:8]: We should add an SQL constraint with `UNIQUE(name, state_id, country_id)`

File: base_location/models/res_company.py
Comment (aitorbouzas) [reply base_location/models/res_company.py]: Yup, it works. Tried several times and it worked for me. Will try more times in case it fails.

**Comment by aitorbouzas** [comment]
@pedrobaeza I think I made all the changes now. Added some tests for the constraints (will wait for codecov's report to add some more and reach 100%)

File: base_location/models/res_city.py
Comment (simahawk) [inline base_location/models/res_city.py:16]: make the error translatable

File: base_location/models/res_city_zip.py
Comment (simahawk) [inline base_location/models/res_city_zip.py:27]: same

File: base_location/models/res_company.py
Comment (simahawk) [inline base_location/models/res_company.py]: `s/./,`

File: base_location/models/res_company.py
Comment (simahawk) [inline base_location/models/res_company.py]: `s/). And [...]/) so we need then to ensure consistency.`

File: base_location/models/res_company.py
Comment (simahawk) [inline base_location/models/res_company.py]: `# As a conclusion`

File: base_location/models/res_company.py
Comment (simahawk) [inline base_location/models/res_company.py]: `. Mixing both approaches...`

File: base_location/models/res_company.py
Comment (simahawk) [inline base_location/models/res_company.py:39]: :heart: this is readonly by default :smile:

File: base_location/models/res_company.py
Comment (simahawk) [inline base_location/models/res_company.py]: `super()` and apply for all occurrences

File: base_location/models/res_partner.py
Comment (simahawk) [inline base_location/models/res_partner.py]:
```python
self.update({
    'zip_id': False,
    'zip': False,
    'city': False,
})
```

File: base_location/models/res_partner.py
Comment (simahawk) [inline base_location/models/res_partner.py:47]: prepare all values and update them at once using `self.update`

File: base_location/models/res_partner.py
Comment (simahawk) [inline base_location/models/res_partner.py]: do not loop twice:
```python
for rec in self:
    if not rec.zip_id:
        continue
[...]
```

File: base_location/models/res_partner.py
Comment (simahawk) [inline base_location/models/res_partner.py:74]: same, use `update(vals)`

File: base_location/tests/test_base_location.py
Comment (simahawk) [inline base_location/tests/test_base_location.py]: 1. `SavepointCase` 
2. `setUp` -> `setUpClass`
3.  disable tracking everywhere
```python
# disable tracking test suite wise
cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
```

File: base_location/tests/test_base_location.py
Comment (simahawk) [inline base_location/tests/test_base_location.py]: `self.env.ref('base.ch').id`

File: base_location/tests/test_base_location.py
Comment (simahawk) [inline base_location/tests/test_base_location.py]: you already have the country recset: `self.env.ref('base.es').write(...)`

File: base_location/tests/test_base_location.py
Comment (simahawk) [inline base_location/tests/test_base_location.py]: its

File: base_location/tests/test_base_location.py
Comment (simahawk) [inline base_location/tests/test_base_location.py]: """Make sure partner data filled accordingly."""

File: base_location/tests/test_base_location.py
Comment (simahawk) [inline base_location/tests/test_base_location.py]: same, you can shorten this docstrings and make them more PEP compliant (not mandatory as we don't enforce this yet in OCA modules)

File: base_location/tests/test_base_location.py
Comment (simahawk) [inline base_location/tests/test_base_location.py]: `self.env.ref`
overall: I'd make all these `ref` usage consistent and just go for `self.env.ref` and then access `.id` when needed

File: base_location/tests/test_base_location.py
Comment (simahawk) [inline base_location/tests/test_base_location.py]: seems like you need these city_zip1/2 in every tests: create them in setUpClass.
Is useless and resource / time consuming when running tests ;)

**Comment by aitorbouzas** [comment]
Fully green. All code covered by tests.

**Review by simahawk — CHANGES_REQUESTED** [review]

File: base_location/models/res_city.py
Comment (pedrobaeza) [reply base_location/models/res_city.py:16]: I would say that it's translatable this way already... Odoo takes care of this. If not, you will have static translations at the server startup.

**Comment by pedrobaeza** [comment]
@aitorbouzas great. I will review later and add migration scripts over your commits.

File: base_location/models/res_company.py
Comment (aitorbouzas) [reply base_location/models/res_company.py]: Sorry, I don't understand what you mean with this comment.

File: base_location/models/res_company.py
Comment (aitorbouzas) [reply base_location/models/res_company.py]: Can you elaborate on this?

File: base_location/models/res_partner.py
Comment (aitorbouzas) [reply base_location/models/res_partner.py]: Does this have an impact in the performance?

File: base_location/models/res_partner.py
Comment (aitorbouzas) [reply base_location/models/res_partner.py]: Well, this is usually running in a recordset with only one record... I understand your code has better performance but also it's less readable.

File: base_location/models/res_company.py
Comment (simahawk) [reply base_location/models/res_company.py]: rephrase like this: "...fields (city, etc..) so we need them to ensure consistency."
(you are already talking about the constrains)

File: base_location/models/res_partner.py
Comment (simahawk) [reply base_location/models/res_partner.py]: yes, you write on it just once. Plus, `update` works on non existing (NewId) items too.

File: base_location/models/res_partner.py
Comment (simahawk) [reply base_location/models/res_partner.py]: Actually, is more readable if you see that you are explicitly discarding records that do not have a zip. With "filtered" you are kind of hiding what your doing (especially to ppl not used to odoo's shortcuts BTW).
Also, using it in loops is a bad practice because in any case you loop twice and you (or whoever takes this code as an example) are going to reuse this approach all around by default "unconsciously".

File: base_location/models/res_company.py
Comment (simahawk) [reply base_location/models/res_company.py]: the dot is wrong, replace it w/ comma :)

**Comment by aitorbouzas** [comment]
@simahawk Thanks for the great review. Changes has been done. :+1:

**Comment by pedrobaeza** [comment]
@aitorbouzas when making changes on a reviewed PR, it's better to add an extra commit, but marked as a fixup, so that the review of these changes doesn't imply to review whole diff again. This is done with:
```
git commit --all --fixup HEAD
```
This way, and before merging, you can squash easily with
```
git rebase -i origin/12.0 --autosquash
```

**Comment by aitorbouzas** [comment]
@pedrobaeza  Oops, i'm such a newbie yet. Thanks for the info.

File: base_location/__manifest__.py
Comment (nikul-serpentcs) [inline base_location/__manifest__.py]: @aitorbouzas change sequence as per [this]( https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L29)

File: base_location/tests/test_base_location.py
Comment (nikul-serpentcs) [inline base_location/tests/test_base_location.py]: @aitorbouzas here `res.country.state` use multiple time, so you should you use global variable.

File: base_location/tests/test_base_location.py
Comment (nikul-serpentcs) [inline base_location/tests/test_base_location.py]: same here

File: base_location/tests/test_base_location.py
Comment (nikul-serpentcs) [inline base_location/tests/test_base_location.py]: same here

File: base_location/tests/test_base_location.py
Comment (nikul-serpentcs) [inline base_location/tests/test_base_location.py]: same here

**Review by nikul-serpentcs — COMMENTED** [review]
Improve Code

File: base_location/tests/test_base_location.py
Comment (aitorbouzas) [reply base_location/tests/test_base_location.py]: I guess you mean to save the ref in a method-scope variable, right?

Comment (simahawk) [reply base_location/tests/test_base_location.py]: @aitorbouzas just to clarify: `s/should/could` => is NOT mandatory.
It's actually a matter of taste + "ease creation" of other records w/out making the code redundant.
For instance: in my personal code I prefer that way as I'm lazy and I hate to copy/paste code :stuck_out_tongue:
but as you've seen I did not complain :wink: 
So all in all: nice to have but up to you.

Comment (aitorbouzas) [reply base_location/tests/test_base_location.py]: @simahawk Oh well, I know that I must have my own criteria for this type of things. Thanks :)

Comment (nikul-serpentcs) [reply base_location/tests/test_base_location.py]: > I guess you mean to save the ref in a method-scope variable, right?

Yes :+1:

**Review by pedrobaeza — APPROVED** [review]
Great work! Maybe you can add another demo data with "United States" or "Spain" that has states, so that the location is complete. Before merging, I want to make a bit of squashing for removing unneeded commits (like "Rename manifest file", etc), but I don't want to give you more work, so I'll do it when other reviewer approve this. We also need migration scripts, that I will add in that step.

**Review by simahawk — APPROVED** [review]
well done! thanks!

**Comment by aitorbouzas** [comment]
I can squash them, but... Can I squash commits from different authors? I don't mind doing it. I can add some demo data too.

**Comment by pedrobaeza** [comment]
I haved squashed commit history and add migration scripts. @aitorbouzas check if everything is correct for merging this.

**Comment by aitorbouzas** [comment]
@pedrobaeza looks good :+1:

**Comment by aitorbouzas** [comment]
With this scripts are ensured the relations between partners/companies and zips?
Edit: And what about if old database has some duplicated zip?

**Comment by pedrobaeza** [comment]
You are right. That one is not preserved right now. I'm going to see if it's possible to be done easily, but it's also non critical, as data are dumped on onchange of the value.

**Comment by pedrobaeza** [comment]
OK, check now that I have covered that part.

**Comment by pedrobaeza** [comment]
The problem with these migration scripts is that right now, we are not testing them anywhere. Maybe they can be integrated on OpenUpgrade tests, but not right now. @StefanRijnhart what do you think about this?

**Comment by aitorbouzas** [comment]
> OK, check now that I have covered that part.

:+1:

**Comment by pedrobaeza** [comment]
Let's go ahead with this for now for being able to migrate other modules depending on this one.

**Comment by emagdalenaC2i** [comment]
Should __openerp__.py be renamed as __manifest__.py?

**Comment by pedrobaeza** [comment]
It's not mandatory for working, but desirable. I'm changing it now.

**Comment by StefanRijnhart** [comment]
> The problem with these migration scripts is that right now, we are not testing them anywhere. Maybe they can be integrated on OpenUpgrade tests, but not right now. @StefanRijnhart what do you think about this?

I'd love to increase the scope of OpenUpgrade testing, but at the moment it's setup to work with a static database with just the core modules installed. I don't know how to set up scalable tests (given the Travis limitation of concurrent jobs) that load OCA modules on the fly and as a test, migrates them to later Odoo versions.

**Pull request merged**
